### PR TITLE
Split detail and record to simplify the code.

### PR DESCRIPTION
**Pull request by rgmf**
### Describe the pull request
- I've split TrackDetailActivity into TrackDetailActivity and TrackRecordActivity.
- I've split StatsFragment into StatsDetailFragment and StatsRecordFragment.
- Both activities share the same ChartFragment because there aren't any difference between detail and record (both use TrackDataHub to feed the charts).

Theses changes simplify the menu logic and record/resume logic. Also, you can create different layouts for recording and detail.

**I've added the datetime of the activity in the Detail layout**. I like to know when I record an activity so i think is useful to put it in the detail. Anyway I can remove it :sweat_smile:

### Link to the the issue
#171 

### License agreement
By opening this pull request, you are providing your contribution under the _Apache License 2.0_ (see [LICENSE.md](LICENSE.md)).

### CODE REVIEW

File: src/main/java/de/dennisguse/opentracks/TrackRecordActivity.java [hunk 1]
```diff
@@ -0,0 +1,456 @@
+/*
+ * Copyright 2008 Google Inc.
```
Comment (dennisguse) [inline src/main/java/de/dennisguse/opentracks/TrackRecordActivity.java:2]: Made my day :)

File: src/main/java/de/dennisguse/opentracks/TrackRecordActivity.java [hunk 2]
```diff
@@ -0,0 +1,456 @@
...
+
+            @NonNull
+            @Override
+            public Fragment getItem(int position) {
+                switch (position) {
+                    case 1:
+                        return ChartFragment.newInstance(false);
+                    case 2:
+                        return ChartFragment.newInstance(true);
+                    default: //0
```
Comment (dennisguse) [inline src/main/java/de/dennisguse/opentracks/TrackRecordActivity.java:218]: Why not make it 0?
And default to throw RuntimeException?

File: src/main/java/de/dennisguse/opentracks/util/StringUtils.java [hunk 3]
```diff
@@ -63,6 +63,11 @@
         ISO_8601_BASE.setTimeZone(TimeZone.getTimeZone("UTC"));
     }
 
+    private static final long SECOND_IN_MILLIS = 1000;
```
Comment (dennisguse) [inline src/main/java/de/dennisguse/opentracks/util/StringUtils.java:66]: Should be in UnitConversions.

File: src/main/java/de/dennisguse/opentracks/util/StringUtils.java [hunk 4]
```diff
@@ -91,6 +96,77 @@ public static String formatDateTimeIso8601(long time_ms) {
...
+        return DateUtils.formatDateTime(context, timeInMillis, flags);
+    }
+
+    /**
+     * Returns the number of days since the epoch (January 01, 1970, 12:00 Midnight UTC).
+     *
+     * @param date date in milliseconds.
+     * @return the number of days in UTC time from the epoch.
+     */
+    public static long getDayNumber(long date) {
```
Comment (dennisguse) [inline src/main/java/de/dennisguse/opentracks/util/StringUtils.java:121]: Why do this computation manually and not using Java's default?
Date expiry = new Date(Long.parseLong(date));

File: src/main/java/de/dennisguse/opentracks/util/StringUtils.java [hunk 5]
```diff
@@ -91,6 +96,77 @@ public static String formatDateTimeIso8601(long time_ms) {
...
+    }
+
+    /**
+     * Converts the datetime into something to display to users.
+     *
+     * @param context      context object.
+     * @param dateInMillis datetime in milliseconds.
+     * @return a string with a friendly user datetime format.
+     */
+    public static String getFriendlyDateTimeString(Context context, long dateInMillis) {
```
Comment (dennisguse) [inline src/main/java/de/dennisguse/opentracks/util/StringUtils.java:156]: Do we have test coverage for this functions?

**Comment by dennisguse** [comment]
@rgmf Pretty nice!

I will just use different names for the StatsFragment as well as RecordFragement (as well as both xml files).
What about StatisticsRecordedFragment and StatisticsRecordingFragment?

File: src/main/java/de/dennisguse/opentracks/TrackRecordActivity.java (hunk 1 above)
Comment (rgmf) [reply src/main/java/de/dennisguse/opentracks/TrackRecordActivity.java:2]: :sweat_smile:

File: src/main/java/de/dennisguse/opentracks/TrackRecordActivity.java (hunk 2 above)
Comment (rgmf) [reply src/main/java/de/dennisguse/opentracks/TrackRecordActivity.java:218]: Better.
I've changed TrackDetailActivity too.

File: src/main/java/de/dennisguse/opentracks/util/StringUtils.java (hunk 3 above)
Comment (rgmf) [reply src/main/java/de/dennisguse/opentracks/util/StringUtils.java:66]: I did all these changes to add datetime in track detail in a human friendly way, but I really don't like so much the way it looks.

Also, layouts will be changed so I'm going to remove all this code ;)

**Comment by rgmf** [comment]
> @rgmf Pretty nice!
> 
> I will just use different names for the StatsFragment as well as RecordFragement (as well as both xml files).
> What about StatisticsRecordedFragment and StatisticsRecordingFragment?

Sounds good and more descriptive :)

**Comment by rgmf** [comment]
Later I'm going to fix a bug I found and push changes ;)

**Comment by rgmf** [comment]
@dennisguse before Resuming feature was added there was a hidden bug (without impact on the final behaviour) but with all these changes I stumple upon it. To reproduce (before):
- Record a new Track.
- While recording go to menu and delete it.
- In the internal code, while service is deleting the track, onResume is called and TrackController is resumed unnecessarily.

Now this bug have a negative impact. This could be fixed with a boolean flag, but I think is a matter of design. I've removed delete option from TrackRecord. If user mistakes and start a track, he can stop an then he'll be in TrackDetail where he can delete the track.

With these changes TrackRecord doesn't need to extend AbstracListActivity.

What do you think? I've created several commits.

**Comment by dennisguse** [comment]
@rgmf Could you rebase the changes on master?
Sadly there are some conflicts (sorry) and I don't know your intentions in detail to be quickly fix it.
Otherwise it looks quite good!

**Comment by rgmf** [comment]
> @rgmf Could you rebase the changes on master?

Of course ;)

> Sadly there are some conflicts (sorry) and I don't know your intentions in detail to be quickly fix it.

No problem, I do it.

**Comment by dennisguse** [comment]
@rgmf My idea was to merge this rather soonish as then I can start working on the elevation gain again and avoid another round of conflicts :)

**Comment by rgmf** [comment]
@dennisguse I've moved your changes on `setHeartRateSensorData`, `setCadenceSensorData` and 
`setSpeedSensorData` from StatsFragment to RecordFragment (the conflict was there).

Rebase done ;)

**Comment by dennisguse** [comment]
Just one more thing: what is the advantage of splitting TrackDetailActivity into two activities?

**Comment by rgmf** [comment]
> Just one more thing: what is the advantage of splitting TrackDetailActivity into two activities?

I think the old Activity focused in couple of things: track stats (A) and track recording (B).

When A -> the activity didn't need TrackRecordingServiceConnection.
When B -> the activity needed TrackRecordingServiceConnection.

When A -> the activity didn't need TrackController and its buttons.
When B -> the activity needed the TrackController and its buttons.

When A -> the activity didn't need flags and status members like recordingTrackPaused, for example.
When B -> the activity needed members to follow the status of recording: recordingTrackId, recordingTrackPaused...

Etc.

Also, every change in the old Activity was more tricky than now. For example, if you wanted to change something with impact only in recording you could to break something on the detail part and the other way around.

Summing up: I think we had an Activity for two things and now we have a more maintainable code.

**Comment by dennisguse** [comment]
I see.

Then we leave it as it is for now.
I was just thinking that we have now animations between these two activities that are not really nice (from the looks, also the state of the tab selector is not transferred or the zoom of the diagrams), but you are right - it makes the code cleaner.
And we can fix on this later ;)

**Comment by rgmf** [comment]
Yes, the animation is weird because it looks a blink. I can work on transition between these two activities. Maybe It can improve.

**Comment by rgmf** [comment]
@dennisguse check my last push out, what do you think something like that?

**Comment by dennisguse** [comment]
Good idea.
Two thoughts about left/right:
1. we would need to apply to this to whole application (especially TrackListActivity)
2. I would go for TrackList -> (left) TrackRecordedActivity -> (left) TrackRecordingActivity and vice versa with right.

What about this: https://developer.android.com/training/transitions/start-activity

**Comment by dennisguse** [comment]
@rgmf Can you also update doc/activities.txt?
It is a PlantUML file describing the flow between activities.

**Comment by rgmf** [comment]
> Good idea.
> Two thoughts about left/right:
> 
>     1. we would need to apply to this to whole application (especially TrackListActivity)
> 
>     2. I would go for TrackList -> (left) TrackRecordedActivity -> (left) TrackRecordingActivity and vice versa with right.

Yes, that's logical.

I've been checking several approaches and I think with a fade animation the transition is more smooth. Tomorrow I'll push the changes.
 
> 
> What about this: https://developer.android.com/training/transitions/start-activity

I did it but those transitions was awful in OpenTracks because every item in the layout moved independently.

Some UI expert will be great here :sweat_smile:

**Comment by rgmf** [comment]
After some tests, I like more a fade in/out for activity transition between recorded and recording. Is a more smooth transition and requires less changes in the code.

Also, I've updated doc/activities.txt.

It's ready for merge.

**Comment by dennisguse** [comment]
Looks pretty awesome!
I love the fade-in/-out animation.

@rgmf For the constraint layout, does the preview work for you in Android Studio?
I also get some warnings (View/TextView not allowed here).

**Comment by rgmf** [comment]
> Looks pretty awesome!
> I love the fade-in/-out animation.

I'm glad you like it :D
 
> @rgmf For the constraint layout, does the preview work for you in Android Studio?
> I also get some warnings (View/TextView not allowed here).

Yes it works well in Android Studio (version I'm using is 3.6.3). No issues, errors or warnings about it. It's strange, do you have any idea what can it is? I'm using Android Studio 3.6.3

I've added tools:text because this help me in design time to see fake data for TextViews where speed, avg. speed, etc, will be shown. Could it be that?

I've tested in Pixel 2 XL and Sony Xperia XZ1 with a screen with only 4.6" and the result its okay.

imagen

imagen

**Comment by dennisguse** [comment]
@rgmf Works now. I needed to run gradle to download constraintlayout dependency.

So, from my side everything looks pretty good!
@rgmf If you believe it is ready, let's merge it.
What would be great, if you could squash the commits by useful chunks of work.
Especially, my two commits :)

**Comment by rgmf** [comment]
> @rgmf If you believe it is ready, let's merge it.
> What would be great, if you could squash the commits by useful chunks of work.
> Especially, my two commits :)

I'll do it ;)
